### PR TITLE
ADM-973 [frontend][backend]: share api

### PR DESCRIPTION
**Pull request by zhou-yinyuan**
## Summary

...

## Before

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## After

_Description_

**Screenshots**
If applicable, add screenshots to help explain behavior of your code.

## Note

_Null_

### CODE REVIEW

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	public static final Long EXPORT_CSV_VALIDITY_TIME = 1000L * 3600 * 24 * 7;
+
+	private static final String CSV_EXTENSION = ".csv";
+
+	private final Gson gson;
+
+	public void createPath(FileType type, String uuid) {
+		Path path = Path.of(BASE_OUTPUT_PATH + SLASH + type.getType() + SLASH + uuid);
+		try {
+			Files.createDirectories(path);
```
[the remark this answers is not in the document]
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/repository/FileRepository.java:56]: done

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+			log.info("Failed create {} directory", path);
+			throw new FileIOException(e);
+		}
+	}
+
+	public <T> T readFileByType(FileType fileType, String uuid, String fileName, Class<T> classType,
+			FilePrefixType fileNamePrefix) {
+		String realFileName = fileNamePrefix.getPrefix() + fileName;
+		File file = new File(getFileName(fileType, uuid, realFileName));
+		if (file.exists()) {
```
[the remark this answers is not in the document]
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/repository/FileRepository.java:72]: done

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+						uuidDirectory, e);
+			}
+		}
+	}
+
+	public List<String> getReportFiles(String uuid) {
+		String fileName = BASE_OUTPUT_PATH + SLASH + REPORT.getPath() + uuid;
+		File folder = new File(fileName);
+
+		if (folder.exists() && folder.isDirectory()) {
```
[the remark this answers is not in the document]
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/repository/FileRepository.java:168]: done

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	public boolean isExpired(long currentTimeStamp, long timeStamp) {
+		return timeStamp < currentTimeStamp - EXPORT_CSV_VALIDITY_TIME;
+	}
+
+	public InputStreamResource readStringFromCsvFile(String uuid, String fileName, FilePrefixType filePrefixType) {
+		File file = new File(getFileName(FileType.CSV, uuid, filePrefixType.getPrefix() + fileName + CSV_EXTENSION));
+		try {
+			InputStream inputStream = new FileInputStream(file);
```
[the remark this answers is not in the document]
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/repository/FileRepository.java:201]: done

File: backend/src/main/java/heartbeat/handler/AsyncReportRequestHandler.java [hunk 5]
```diff
@@ -33,4 +38,17 @@
 		deleteExpireFileByType(REPORT, currentTimeStamp, directory);
 	}
 
+	public List<String> getReportFiles(String uuid) {
+		Path targetPath = new File(OUTPUT_FILE_PATH).toPath().normalize();
+		String fileName = targetPath + SLASH + REPORT.getPath() + uuid;
+		File folder = new File(fileName);
+
+		if (folder.exists() && folder.isDirectory()) {
```
[the remark this answers is not in the document]
Comment (zhou-yinyuan) [reply backend/src/main/java/heartbeat/handler/AsyncReportRequestHandler.java]: done

**Review by guzhongren — APPROVED** [review]

File: backend/src/main/java/heartbeat/controller/report/ReportController.java [hunk 6]
```diff
@@ -43,42 +41,53 @@ public class ReportController {
...
+		log.info("Start to generate report_reportId: {}", uuid);
+		return generateReporterService.getComposedReportResponse(uuid, startTime, endTime);
 	}
 
-	@PostMapping
-	public ResponseEntity<CallbackResponse> generateReport(@RequestBody GenerateReportRequest request) {
+	@GetMapping("/{uuid}")
+	public ShareApiDetailsResponse getShareDetails(@PathVariable String uuid) {
+		log.info("start to get share details, uuid: {}", uuid);
+		return reportService.getShareReportInfo(uuid);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/ReportController.java:69]: add log to record the request finished.

same as others

File: backend/src/main/java/heartbeat/controller/report/ReportController.java [hunk 7]
```diff
@@ -43,42 +41,53 @@ public class ReportController {
...
+	}
+
+	@PostMapping("/{uuid}")
+	public ResponseEntity<CallbackResponse> generateReport(@PathVariable String uuid,
+			@RequestBody GenerateReportRequest request) {
 		log.info("Start to generate report");
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/ReportController.java:75]: record the parameter(uuid) at the start and end of the log

File: backend/src/main/java/heartbeat/controller/report/ReportController.java [hunk 8]
```diff
@@ -43,42 +41,53 @@ public class ReportController {
...
+				TimeUtil.convertToUserSimpleISOFormat(Long.parseLong(request.getEndTime()),
+						request.getTimezoneByZoneId()));
 		log.info("Successfully generate report");
 		return ResponseEntity.status(HttpStatus.ACCEPTED)
 			.body(CallbackResponse.builder().callbackUrl(callbackUrl).interval(interval).build());
 	}
 
+	@PostMapping
+	public UuidResponse generateUUID() {
+		return reportService.generateReportId();
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/ReportController.java:90]: add log

File: backend/src/main/java/heartbeat/controller/report/dto/response/ShareApiDetailsResponse.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+
+import java.util.List;
+
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+public class ShareApiDetailsResponse {
+
+	private List<String> reportUrls;
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/dto/response/ShareApiDetailsResponse.java:16]: rename to `reportURLs`

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 10]
```diff
@@ -0,0 +1,255 @@
...
+	public void createPath(FileType type, String uuid) {
+		isCorrectFilePath(uuid);
+
+		Path path = Path.of(BASE_OUTPUT_PATH + SLASH + type.getType() + SLASH + uuid);
+		try {
+			Files.createDirectories(path);
+			log.info("Successfully create {} directory", path);
+		}
+		catch (IOException e) {
+			log.info("Failed create {} directory", path);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:61]: Failed to create ....

Comment (guzhongren) [reply backend/src/main/java/heartbeat/repository/FileRepository.java:61]: Log.error

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 11]
```diff
@@ -0,0 +1,255 @@
...
+		File file = new File(getFileName(fileType, uuid, realFileName));
+		if (file.toPath().normalize().startsWith(NORMALIZE_BASE_OUTPUT_PATH) && file.exists()) {
+			try (JsonReader reader = new JsonReader(new FileReader(file))) {
+				T result = gson.fromJson(reader, classType);
+				log.info("Successfully read file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid,
+						realFileName);
+				return result;
+			}
+			catch (Exception e) {
+				log.error("Failed read file type: {}, uuid: {}, file name: {}, reason: {}", fileType.getType(), uuid,
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:81]: Failed to read.....

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 12]
```diff
@@ -0,0 +1,255 @@
...
+						realFileName, e);
+				throw new GenerateReportException(
+						"Failed read file " + fileType.getType() + " " + uuid + " " + realFileName);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:84]: Failed to read....

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+		String realFileName = fileNamePrefix.getPrefix() + fileName;
+		String path = getFileName(fileType, uuid, realFileName);
+
+		log.info("Start to remove file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid, fileName);
+		try {
+			Files.deleteIfExists(Path.of(path));
+			log.info("Successfully remove file type: {}, file name: {}", fileType.getType(), fileName);
+		}
+		catch (Exception e) {
+			log.info("Failed remove file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid, realFileName);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:128]: Failed to remove
log.error


Please check all your error log

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+			throw new GenerateReportException(
+					"Failed remove " + fileType.getType() + ", uuid: " + uuid + " file with file:" + fileName);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:130]: Failed to remove

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 15]
```diff
@@ -0,0 +1,255 @@
...
+			throw new FileIOException(e);
+		}
+	}
+
+	private void createFileHandler(FileType fileType, String uuid, String fileName, FilePrefixType fileNamePrefix,
+			Consumer<String> handler) {
+		createPath(fileType, uuid);
+		String realBaseFileName = fileNamePrefix.getPrefix() + fileName;
+		String realFileName = getFileName(fileType, uuid, realBaseFileName);
+		log.info("Start to write file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid, realFileName);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:215]: logs should be a pair, there are `start and end` for the log

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+		String tmpFileName = realFileName + SUFFIX_TMP;
+
+		try (FileWriter writer = new FileWriter(tmpFileName)) {
+			writer.write(json);
+			Files.move(Path.of(tmpFileName), Path.of(realFileName), StandardCopyOption.ATOMIC_MOVE);
+			log.info("Successfully write file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid,
+					realFileName);
+		}
+		catch (Exception e) {
+			log.error("Failed write file type: {}, uuid: {}, file name: {}, reason: {}", fileType.getType(), uuid,
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:231]: Failed to write....

**Review by guzhongren — CHANGES_REQUESTED** [review]

File: backend/src/main/java/heartbeat/controller/report/ReportController.java [hunk 17]
```diff
@@ -60,34 +60,42 @@ public ReportResponse generateReport(@PathVariable String uuid,
 			@Schema(type = "string", example = "20240310", pattern = "^[0-9]{8}$") @Parameter String startTime,
 			@Schema(type = "string", example = "20240409", pattern = "^[0-9]{8}$") @Parameter String endTime) {
 		log.info("Start to generate report_reportId: {}", uuid);
-		return generateReporterService.getComposedReportResponse(uuid, startTime, endTime);
+		ReportResponse composedReportResponse = generateReporterService.getComposedReportResponse(uuid, startTime,
+				endTime);
+		log.info("successfully to generate report_reportId: {}", uuid);
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/ReportController.java:65]: Successfully generate

Comment (guzhongren) [reply backend/src/main/java/heartbeat/controller/report/ReportController.java:65]: same as others..
Keep first character as Upcase

File: backend/src/main/java/heartbeat/controller/report/ReportController.java [hunk 18]
```diff
@@ -60,34 +60,42 @@ public ReportResponse generateReport(@PathVariable String uuid,
...
 		return ResponseEntity.status(HttpStatus.ACCEPTED)
 			.body(CallbackResponse.builder().callbackUrl(callbackUrl).interval(interval).build());
 	}
 
 	@PostMapping
 	public UuidResponse generateUUID() {
-		return reportService.generateReportId();
+		log.info("start to generate uuid");
+		UuidResponse uuidResponse = reportService.generateReportId();
+		log.info("successfully to generate uuid");
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/controller/report/ReportController.java:97]: log the `uuid`

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 19]
```diff
@@ -73,15 +73,15 @@ public <T> T readFileByType(FileType fileType, String uuid, String fileName, Cla
 		if (file.toPath().normalize().startsWith(NORMALIZE_BASE_OUTPUT_PATH) && file.exists()) {
 			try (JsonReader reader = new JsonReader(new FileReader(file))) {
 				T result = gson.fromJson(reader, classType);
-				log.info("Successfully read file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid,
+				log.info("Successfully to read file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid,
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:76]: No `to`

File: backend/src/main/java/heartbeat/repository/FileRepository.java [hunk 20]
```diff
@@ -228,16 +231,16 @@ private void createNormalFileHandler(FileType fileType, String uuid, String json
...
-			throw new GenerateReportException("Failed write " + fileType.getType() + " " + realFileName);
+			throw new GenerateReportException("Failed to write " + fileType.getType() + " " + realFileName);
 		}
 	}
 
 	private void createCSVFileHandler(FileType fileType, String uuid, String[][] json, String realFileName) {
 		try (CSVWriter writer = new CSVWriter(new FileWriter(realFileName))) {
 			writer.writeAll(Arrays.asList(json));
-			log.info("Successfully write file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid,
+			log.info("Successfully to write file type: {}, uuid: {}, file name: {}", fileType.getType(), uuid,
```
Comment (guzhongren) [inline backend/src/main/java/heartbeat/repository/FileRepository.java:243]: same issue

**Review by guzhongren — APPROVED** [review]
@zhou-yinyuan small change 
Approval it, merge it after changing with the comments

**Pull request merged**